botocore/utils.py

```diff
@@ -794,5 +800,138 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
+
+    def redirect_from_error(self, request_dict, response, **kwargs):
+        """
+        An S3 request sent to the wrong region will return an error that
+        contains the endpoint the request should be sent to. This handler
+        will add the redirect information to the signing context and then
+        redirect the request.
+        """
+        if response is None:
+            # This could be none if there was a ConnectionError or other
+            # transport error.
+            return
+
+        error = response[1].get('Error', {})
+        if error.get('Code') != 'PermanentRedirect':
+            return
+
+        bucket = error['Bucket']
+        client_region = request_dict['context'].get('client_region')
+        new_region = self.get_bucket_region(response)
+
+        if new_region is None:
+            logger.debug(
+                "S3 client configured for region %s but the bucket %s is not "
+                "in that region and the proper region could not be "
+                "automatically determined." % (client_region, bucket))
+            return
+
+        logger.debug(
+            "S3 client configured for region %s but the bucket %s is in region"
+            " %s; Please configure the proper region to avoid multiple "
+            "unnecessary redirects and signing attempts." % (
+                client_region, bucket, new_region))
+
+        endpoint = self._endpoint_resolver.resolve('s3', new_region)
+        endpoint = endpoint['endpoint_url']
+
+        signing_context = {
+            'region': new_region,
+            'bucket': bucket,
+            'endpoint': endpoint
+        }
+        request_dict['context']['signing'] = signing_context
+
+        self._cache.set(bucket, signing_context)
+        self.set_request_url(request_dict, request_dict['context'])
+
+        # Return 0 so it doesn't wait to retry
+        return 0
+
+    def get_bucket_region(self, response):
+        """
+        There are multiple potential sources for the new region to redirect to,
+        but they aren't all universally available for use. This will try to
+        find region from response elements, but will fall back to calling
+        HEAD on the bucket if all else fails.
+
+        :param response: A response representing a service request that failed
+            due to incorrect region configuration.
+        """
+        # First try to source the region from the headers.
+        service_response = response[1]
+        response_headers = service_response['ResponseMetadata']['HTTPHeaders']
+        if 'x-amz-bucket-region' in response_headers:
+            return response_headers['x-amz-bucket-region']
+
+        # Next, check the error body
+        region = service_response.get('Error', {}).get('Region')
+        if region:
+            return region
+
+        # Finally, HEAD the bucket. No other choice sadly.
+        bucket = service_response.get('Error', {}).get('Bucket')
+        self.unregister()
```

Comment:
I would be very wary of doing unregister and registering mid-request for concurrencyu. If a client is being shared what happens if you unregistered the request mid another thread using the client to send a request that has the incorrect bucket region? It seems that the redirection functionality may have a chance of slipping through the cracks here if we are possibly registering and unregistering these events.